test/extensions.js

```diff
@@ -0,0 +1,133 @@
+var tape = require('tape')
+var memdb = require('memdb')
+var signatures = require('sodium-signatures')
+var hypercore = require('../')
+
+tape('use extension', function (t) {
+  var core = create()
+  var remote = create()
+
+  var ping = core.use('ping')
+  remote.use('ping')
+
+  var coreStream = core.createPeerStream()
+  var remoteStream = remote.createPeerStream()
+
+  remoteStream.pipe(coreStream).pipe(remoteStream)
+
+  coreStream.on('handshake', function () {
+    if (coreStream.remoteSupports(ping)) {
+      coreStream.send(ping, new Buffer([0])) // send ping
+
+      remoteStream.on(ping, function (buf) {
+        if (buf[0] === 0) remoteStream.send(ping, new Buffer([1])) // send pong
+      })
+
+      coreStream.on(ping, function (buf) {
+        if (buf[0] === 1) {
+          t.same(buf, new Buffer([1]), 'pong recieved')
+          t.end()
+        }
+      })
+    }
+  })
+})
+
+tape('no extension support', function (t) {
+  var core = create()
+  var remote = create()
+
+  var ping = core.use('ping')
+
+  var coreStream = core.createPeerStream()
+  var remoteStream = remote.createPeerStream()
+
+  remoteStream.pipe(coreStream).pipe(remoteStream)
+
+  coreStream.on('handshake', function () {
+    t.ok(!coreStream.remoteSupports(ping), 'peer does not support extension')
+    t.end()
+  })
+})
+
+tape('multiple extensions', function (t) {
+  var core = create()
+  var remote = create()
+
+  core.use('snakes')
+  core.use('ping')
+  remote.use('ping')
+
+  var coreStream = core.createPeerStream()
+  var remoteStream = remote.createPeerStream()
+
+  remoteStream.pipe(coreStream).pipe(remoteStream)
+
+  coreStream.on('handshake', function () {
+    t.same(coreStream.listAvailable(), [ 'ping' ], 'multiple extension works')
+    t.end()
+  })
+})
+
+tape('multiple extension message routing', function (t) {
+  var core = create()
+  var remote = create()
+
+  core.use('snakes')
+  var ping = core.use('ping')
+  remote.use('ping')
+
+  var coreStream = core.createPeerStream()
+  var remoteStream = remote.createPeerStream()
+
+  remoteStream.pipe(coreStream).pipe(remoteStream)
+
+  coreStream.on('handshake', function () {
+    coreStream.send(ping, new Buffer([0]))
+
+    remoteStream.on(ping, function (buf) {
```

Comment:
I think that this test is "cheating" a little bit because we're using the `ping` variable to listen on the remote stream, but we got the value of `ping` from the core stream. IRL, we wouldn't be able to do this since they would be running on different machines.